### ISSUE TITLE
Login to registries / Error response from daemon: toomanyrequests: You have reached your pull rate limit.

### ISSUE THREAD

**Issue by Nils98Ar** [SovereignCloudStack/k8s-cluster-api-provider#414]
Hi,

when upgrading some clusters we sometimes have problems with container registry rate limits. The current workaround is to wait for 6 hours ;)

Is there a way to inject credentials for registry login e.g. for docker hub?

Best regards
Nils

**Comment by Nils98Ar** [comment]
Maybe we can just provide the credentials like this? https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials

**Comment by jschoone** [comment]
Hi @Nils98Ar and sorry for the late response.
My approach would to replace all `docker.io` entries with another valid registry.
For example calico host all their images at `quay.io`
I just did a `sed 's/docker.io/quay.io/g'` on `terraform/files/kubernetes-manifests.d/calico.yaml` and tested it, which worked. You will find this in [this branch](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/tree/chore/replace_dockerio).

Another approach could be using a mirror on front of the registry, which we want to provide soon. But I think this is another topic and it is still a good idea to find other registries than docker.io.

@matofeder and @chess-knight what do you think is the best approach?
1. Should we hard core the registries in the manifest?
2. Should we patch the manifest during deployment?

**Comment by Nils98Ar** [comment]
No problem for now I do only two upgrades per day and check the quotas before starting ;)

In my opinion the manifests need to be patched when pulling via curl from github or when applying.

**Comment by matofeder** [comment]
@Nils98Ar thanks for bringing this up. We already discussed this topic in one of the SCS team meeting and we agreed that it is a good idea to use `registry.scs.community` as a mirror for `docker.io` and maybe also other container sources to avoid pull rate limit issues. 

> It turned out that creating the secret is not enough. It needs to be used too :)

Yes, you should adjust "affected" manifests where you want to use credentials for docker.io and reference created secret in  `imagePullSecrets`.

Another workaround could be to set your docker.io credentials globally in CRI configuration file. SCS k8s clusters use `containerd` as CRI. See related [docs](https://github.com/containerd/containerd/blob/main/docs/cri/registry.md#configure-registry-credentials), but this may require containerd restart, which is not optimal, and of course, you should have access to k8s nodes.  This change is global, i.e. affects all pulls from docker.io. 
 
> @matofeder and @chess-knight what do you think is the best approach?
> 
>     1. Should we hard core the registries in the manifest?
> 
>     2. Should we patch the manifest during deployment?
> 
>     3. Other ideas

IMO this requires further discussion. I created a separate issue https://github.com/SovereignCloudStack/k8s-cluster-api-provider/issues/417, with some open questions where we could propose a proper solution.
